### ISSUE TITLE
Bug: Multi-GPU DDP training hangs at specific epoch on v8.4.60 but works perfectly on v8.4.9

### ISSUE THREAD

**Issue by xusuyong** [ultralytics/ultralytics#24675]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Multi-GPU

### Bug

Description

When training yolov8s using Multi-GPU DDP (2x RTX 3090), the training process completely hangs at a specific epoch without throwing any errors or exceptions.

    When epochs=120 and batch=120, it consistently hangs at Epoch 30.

    When epochs=30 and batch=120, it consistently hangs at Epoch 27.

    If I change batch=100, it runs smoothly to completion.

    Single-GPU training works completely fine under all configurations.

Crucial Version Clue:
This is a regression bug related to recent updates.

    ultralytics==8.4.9: Works perfectly (Runs to completion with batch=120 on dual GPUs).

<img width="1156" height="1041" alt="Image" src="https://github.com/user-attachments/assets/cfa3f26a-6cff-457b-84f0-4575c0838ac6" />

pip install -i https://pypi.tuna.tsinghua.edu.cn/simple ultralytics -U -q

    ultralytics==8.4.60: Reproducibly hangs as described above. Downgrading back to 8.4.9 immediately resolves the issue.

<img width="1102" height="580" alt="Image" src="https://github.com/user-attachments/assets/3da25119-8184-4daa-b4c4-083d0d950e79" />

### Environment

vision@vision-System-Product-Name /m/w/x/p/m/9/yolov8 (main) [1]> yolo check                                                                                                                                                                                        (pt29)
Ultralytics 8.4.60 🚀 Python-3.11.11 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3090, 24124MiB)
Setup complete ✅ (56 CPUs, 31.2 GB RAM, 24.5/152.6 GB disk)

OS                     Linux-6.8.0-111-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.11.11
Install                pip
Path                   /home/vision/anaconda3/envs/pt29/lib/python3.11/site-packages/ultralytics
RAM                    31.20 GB
Disk                   24.5/152.6 GB
CPU                    Intel Xeon E5-2680 v4 2.40GHz
CPU count              56
GPU                    NVIDIA GeForce RTX 3090, 24124MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 2.3.4>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
import sys
import os
from ultralytics import YOLO

if __name__ == "__main__":
    data_cfg_path = f"../cfg/SOD123-2D-RGB_v1_.yaml"
    model = YOLO(
        "../model/yolov8s.pt",
        task="detect",
    )

    epochs = 30  # 训练轮数

    model.train(
        data=data_cfg_path,  # 你的数据配置文件路径
        epochs=epochs,  # 训练轮数
        imgsz=640,  # 输入图像尺寸
        batch=120,
        name="train_" + dataset_name,  # 子目录名
        mosaic=0.8,  # 默认1，拼图增强的概率。 仅当允许图像中存在部分遮挡物体或多物体共存，且这些情况不会改变标签值时(类别)，才使用拼图增强。mosaic=1.0 表示所有图像都将进行拼图增强，
        workers=3,
        device=[0, 1],
    )
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xusuyong, thank you for reporting this multi-GPU DDP regression in Ultralytics 🚀. This is an automated response to help us triage your issue, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this is a 🐛 Bug Report, thank you for already including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) and detailed environment information 🙌. If possible, please also add any full console logs from the hang point and confirm whether the same behavior persists after upgrading to the latest package version below.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
thanks for the detailed report and the version range, that helps a lot. the multi-gpu ddp wrapper config changed in `8.4.49`, so before fixing we'd like to confirm that's where the regression starts.

could you run one more check: install `8.4.48` and repeat the same run with `batch=120` on both gpus?

```bash
pip install ultralytics==8.4.48
```

if `8.4.48` completes and `8.4.49` hangs, that pins the regression and we can fix it directly.

from your screenshots the hang lands at epoch 27, and in the working `8.4.9` run that same epoch 27 contains a batch with no objects (`Instances 0`, with a cls_loss spike). on a small or sparse dataset a per-gpu sub-batch can end up with no objects, which is a likely trigger for the ddp stall, so this lines up.

if it still hangs on `8.4.48`, a stack trace of both training processes at the hang point would tell us where they are stuck. either option works:

- [py-spy](https://github.com/benfred/py-spy) (`pip install py-spy`), then `py-spy dump --pid <pid>` for each of the two training processes, no code changes needed.
- built-in `faulthandler`: add `import faulthandler, signal; faulthandler.register(signal.SIGUSR1)` at the top of your script, then `kill -USR1 <pid>` when it hangs to print the traceback without killing the process.
- or rerun with `NCCL_DEBUG=INFO` and share the tail of the output.

**Comment by xusuyong** [comment]
@raimbekovm 
Confirmed:

- `ultralytics==8.4.48`: ✅ Completed successfully with `batch=120`, `epochs=30`, `device=[0, 1]`
- `ultralytics==8.4.49`: ❌ Hangs at epoch 27, same as 8.4.60

So the regression is confirmed in **8.4.49**. Hope this helps pin down the fix!

**Comment by filaPro** [comment]
8.4.58 exact same issue exists

**Comment by glenn-jocher** [comment]
Thanks, this is helpful — the bisect to `8.4.49` confirms the regression point, and `@filaPro` reporting the same behavior on `8.4.58` suggests the same DDP path; the proposed fix is in [PR #24676](https://github.com/ultralytics/ultralytics/pull/24676), so if either of you can test that patch on your 2-GPU setup and confirm whether `batch=120` completes, that would help us validate it quickly.

**Comment by xusuyong** [comment]
@glenn-jocher 
Tested locally by manually applying the same logic (find_unused_parameters=not self.args.compile). Training completes successfully with batch=120, epochs=30, device=[0,1] — no hang at epoch 27. PR #24676 should be the correct fix. ✅

**Comment by raimbekovm** [comment]
@xusuyong thanks for the report
